tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc

```diff
@@ -143,17 +185,33 @@ Status Visitor::HandleBatchNormTraining(HloInstruction* batch_norm) {
           computation_->AddInstruction(HloInstruction::CreateBroadcast(
               variance_plus_epsilon->shape(), epsilon, {}))));
 
-  // Repackage the results.
-  std::unique_ptr<HloInstruction> new_tuple = HloInstruction::CreateTuple({
-      computation_->AddInstruction(HloInstruction::CreateGetTupleElement(
-          libcall->shape().tuple_shapes(0), libcall, 0)),
+  HloInstruction* new_tuple =
+      computation_->AddInstruction(HloInstruction::CreateTuple({
+          computation_->AddInstruction(HloInstruction::CreateGetTupleElement(
+              libcall->shape().tuple_shapes(0), libcall, 0)),
+          computation_->AddInstruction(HloInstruction::CreateGetTupleElement(
+              libcall->shape().tuple_shapes(1), libcall, 1)),
+          variance,
+      }));
+
+  HloInstruction* new_gte =
       computation_->AddInstruction(HloInstruction::CreateGetTupleElement(
-          libcall->shape().tuple_shapes(1), libcall, 1)),
-      variance,
-  });
+          new_tuple->shape().tuple_shapes(0), new_tuple, 0));
+
+  if (IsF32BatchNormWithFP16Inputs(batch_norm)) {
+    new_gte = AddConvert(new_gte, F32);
+  }
+  // Repackage the results. Athough this tuple is redundant when convert is not
+  // inserted, TupleSimplifier eliminates the Tuple eventually
+  std::unique_ptr<HloInstruction> replacing_tuple = HloInstruction::CreateTuple(
+      {new_gte,
+       computation_->AddInstruction(HloInstruction::CreateGetTupleElement(
+           new_tuple->shape().tuple_shapes(1), new_tuple, 1)),
+       computation_->AddInstruction(HloInstruction::CreateGetTupleElement(
+           new_tuple->shape().tuple_shapes(1), new_tuple, 2))});
```

Comment:
Is all of this repackaging necessary or can we directly create the final tuple using GTE's on `libcall`, `variance` etc.?

Reply:
Same comment for `HandleBatchNormGrad`.

Reply:
You mean `new_tuple` could have been bypassed..right? Agreed! I initially had it here to signify that the instructions between `new_tuple` and `replacing_tuple` correspond to handling F16 batchnorm. I agree that it doesn't improve anything in terms of readability or making the code more understandable. Note that `HandleBatchNormGrad` doesn't have any additional tuples. The comments `// Repackage the results. Athough ...` implies that instead of adding more code here to not have another GTE for F32, I chose to have one and let the optimizer eliminate the redundant tuple.

Reply:
I will modify `HandleBatchNormTraining` but I don't think `HandleBatchNormGrad` needs changes unless I am misinterpreting your comment :).